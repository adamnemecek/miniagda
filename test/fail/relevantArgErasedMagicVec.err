MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "relevantArgErasedMagicVec.ma" ---
--- scope checking ---
--- type checking ---
type  Sigma : ^(A : Set) -> ^(B : A -> Set) -> Set
term  pair : .[A : Set] -> .[B : A -> Set] -> ^(fst : A) -> ^(snd : B fst) -> < pair fst snd : Sigma A B >
term  fst : .[A : Set] -> .[B : A -> Set] -> (pair : Sigma A B) -> A
{ fst [A] [B] (pair #fst #snd) = #fst
}
term  snd : .[A : Set] -> .[B : A -> Set] -> (pair : Sigma A B) -> B (fst [A] [B] pair)
{ snd [A] [B] (pair #fst #snd) = #snd
}
type  Nat : Set
term  zero : < zero : Nat >
term  succ : ^(y0 : Nat) -> < succ y0 : Nat >
type  Empty : Set
term  magic : .[A : Set] -> .[p : Empty] -> A
{}
type  Unit : Set
term  unit : < unit : Unit >
type  Vec : .[A : Set] -> (n : Nat) -> Set
error during typechecking:
Vec
/// clause 2
/// right hand side
/// checkExpr 2 |- Sigma A (\ z -> Vec A n) : Set
/// inferExpr' Sigma A (\ z -> Vec A n)
/// inferExpr' Sigma A
/// inferExpr' A
/// inferExpr: variable A : Set may not occur
/// , because it is marked as erased
