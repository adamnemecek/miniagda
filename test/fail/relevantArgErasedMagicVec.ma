-- proof irrelevance via polymorphism

data Sigma (A : Set) (B : A -> Set) : Set
{ pair : (fst : A) -> (snd : B fst) -> Sigma A B
}
fields fst, snd

data Nat : Set
{ zero : Nat
; succ : Nat -> Nat
}

data Empty : Set
{
}

-- magic = abort  does not need the inhabitant p : Empty
fun magic : [A : Set] -> [p : Empty] -> A
{ 
}

data Unit : Set
{ unit : Unit
}

fun Vec : [A : Set] -> (n : Nat) -> Set
{ Vec A zero     = Empty
; Vec A (succ n) = Sigma A (\ z -> Vec A n)
}

fun Leq : (n : Nat) -> (m : Nat) -> Set
{ Leq  zero     m        =  Unit
; Leq (succ n)  zero     =  Empty
; Leq (succ n) (succ m)  =  Leq n m
}
let Lt : (n : Nat) -> (m : Nat) -> Set
       = \ n -> \ m -> Leq (succ n) m

fun lookup : [A : Set] -> (n : Nat) -> (m : Nat) -> [Lt m n] -> Vec A n -> A
{ lookup A  zero    m        p v = magic A p
; lookup A (succ n) zero     p v = fst v -- fst A (\ z -> Vec A n) v
; lookup A (succ n) (succ m) p v = lookup A n m p <| snd v -- (snd A (\ z -> Vec A n) v)
}

