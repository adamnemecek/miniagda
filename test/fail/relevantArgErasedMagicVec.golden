MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "relevantArgErasedMagicVec.ma" ---
--- scope checking ---
--- type checking ---
type  Sigma : ^(A : Set) -> ^(B : A -> Set) -> Set
term  Sigma.pair : .[A : Set] -> .[B : A -> Set] -> ^(fst : A) -> ^(snd : B fst) -> < Sigma.pair fst snd : Sigma A B >
term  fst : .[A : Set] -> .[B : A -> Set] -> (pair : Sigma A B) -> A
{ fst [A] [B] (Sigma.pair #fst #snd) = #fst
}
term  snd : .[A : Set] -> .[B : A -> Set] -> (pair : Sigma A B) -> B (fst [A] [B] pair)
{ snd [A] [B] (Sigma.pair #fst #snd) = #snd
}
type  Nat : Set
term  Nat.zero : < Nat.zero : Nat >
term  Nat.succ : ^(y0 : Nat) -> < Nat.succ y0 : Nat >
type  Empty : Set
term  magic : .[A : Set] -> .[p : Empty] -> A
{}
type  Unit : Set
term  Unit.unit : < Unit.unit : Unit >
type  Vec : .[A : Set] -> (n : Nat) -> Set
error during typechecking:
Vec
/// clause 2
/// right hand side
/// checkExpr 2 |- Sigma A (\ z -> Vec A n) : Set
/// inferExpr' Sigma A (\ z -> Vec A n)
/// inferExpr' Sigma A
/// checkApp (^(A : Set) -> ^(B : A -> Set) -> Set) eliminated by A
/// inferExpr' A
/// inferExpr: variable A : Set may not occur
/// , because it is marked as erased
